iree/compiler/Dialect/Stream/Transforms/ScheduleAllocation.cpp

```diff
@@ -0,0 +1,1440 @@
+    auto resourceAccess = IREE::Stream::ResourceAccessBitfieldAttr::get(
+        builder.getContext(), IREE::Stream::ResourceAccessBitfield::Write);
+    newResources.push_back(resourceRange.resource);
+    newResourceSizes.push_back(resourceRange.resourceSize);
+    newResourceOffsets.push_back(resourceOffset);
+    newResourceLengths.push_back(resourceLength);
+    newResourceAccesses.push_back(resourceAccess);
+  }
+
+  builder.create<IREE::Stream::CmdDispatchOp>(
+      asyncOp.getLoc(), asyncOp.workgroup_count(), asyncOp.entry_point(),
+      newOperands, newResources, newResourceSizes, newResourceOffsets,
+      newResourceLengths, builder.getArrayAttr(newResourceAccesses));
+  asyncOp.erase();
+  return success();
+}
+
+static LogicalResult applyAsyncAllocations(Region &region,
+                                           AllocationScope &scope);
+
+static LogicalResult applyAsyncConcurrentOp(
+    IREE::Stream::AsyncConcurrentOp asyncOp, AllocationScope &scope,
+    OpBuilder builder) {
+  // Remove operands from the yield now that we aren't returning anything.
+  // Must do this before we recurse so that the ops we are transforming have no
+  // uses.
+  auto yieldOp =
+      cast<IREE::Stream::YieldOp>(asyncOp.body().front().getTerminator());
+  yieldOp->eraseOperands(0, yieldOp.getNumOperands());
+
+  // Recurse into the wave ops to process the async ops within.
+  // They'll be mapped to resources in the parent scope - we must fix this up!
+  if (failed(applyAsyncAllocations(asyncOp.body(), scope))) {
+    return failure();
+  }
+
+  // Explicit variant has no args/operands
+  auto &block = asyncOp.body().front();
+  block.eraseArguments([&](auto arg) { return true; });
+
+  // Rewrite wave op to remove results.
+  auto newOp = builder.create<IREE::Stream::CmdConcurrentOp>(asyncOp.getLoc());
+  newOp.body().takeBody(asyncOp.body());
+  asyncOp.erase();
+  return success();
+}
+
+// Converts async operations to explicit commands using the allocation mappings
+// in |scope|. Upon successful return the region should have no more defined
+// values.
+static LogicalResult applyAsyncAllocations(Region &region,
+                                           AllocationScope &scope) {
+  // Walk the ops backwards so that we can delete them, freeing uses so that
+  // producers can be deleted in turn.
+  auto &block = region.getBlocks().front();
+  auto ops = llvm::to_vector<4>(llvm::map_range(
+      llvm::reverse(block), [&](Operation &op) { return &op; }));
+  for (auto *op : ops) {
+    if (op->hasTrait<OpTrait::IsTerminator>()) continue;
+    if (failed(TypeSwitch<Operation *, LogicalResult>(op)
+                   .Case([&](IREE::Stream::ResourceSubviewOp op) {
+                     return applyResourceSubviewOp(op, scope, OpBuilder(op));
+                   })
+                   .Case([&](IREE::Stream::AsyncAllocaOp op) {
+                     return applyAsyncAllocaOp(op, scope, OpBuilder(op));
+                   })
```

Comment:
(no action needed)

ew indentation x_x

the Google c++ style guide also has some preferences against default capture by reference: https://google.github.io/styleguide/cppguide.html#Lambda_expressions, but I think these uses are fine

also https://llvm.org/docs/CodingStandards.html#format-lambdas-like-blocks-of-code

Reply:
blame clang-format for this - not much one can do that doesn't just make it worse (adding clang-format off and stuff) :(